File: py/generate.py
"""Prefill / decode loop with streaming and optional padded batching."""

from __future__ import annotations

import sys
import time
from collections.abc import Iterator
from pathlib import Path

import numpy as np

from .cache import KVCache
from .model import LlamaModel
from .sampler import sample_token
from .tokenizer import StreamDecoder, Tokenizer, apply_chat_template


def _as_batch(token_lists: list[list[int]], pad_id: int) -> tuple[np.ndarray, np.ndarray]:
    lengths = np.array([len(t) for t in token_lists], dtype=np.int32)
    slen = int(lengths.max()) if len(lengths) else 0
    batch = np.full((len(token_lists), slen), pad_id, dtype=np.int32)
    for i, ids in enumerate(token_lists):
        batch[i, : len(ids)] = np.asarray(ids, dtype=np.int32)
    return batch, lengths


def generate(
    model: LlamaModel,
    tokenizer: Tokenizer,
    prompt_ids: list[int],
    max_tokens: int = 128,
    temperature: float = 0.8,
    top_k: int = 0,
    top_p: float = 0.9,
    cache: KVCache | None = None,
    rng: np.random.Generator | None = None,
    stop_ids: set[int] | None = None,
) -> Iterator[int]:
    """Yield generated token ids. Stops on EOS / END without yielding them."""
    if cache is None:
        cache = model.new_cache(batch=1)
    if stop_ids is None:
        stop_ids = tokenizer.stop_ids
    prompt = np.asarray(prompt_ids, dtype=np.int32)[None, :]
    logits = model.forward(prompt, cache)
    for _ in range(max_tokens):
        tid = sample_token(logits[0], temperature, top_k, top_p, rng)
        if tid in stop_ids:
            return
        yield tid
        logits = model.forward(np.array([[tid]], dtype=np.int32), cache)


def generate_text(
    model: LlamaModel,
    tokenizer: Tokenizer,
    prompt: str,
    max_tokens: int = 128,
    temperature: float = 0.8,
    top_k: int = 0,
    top_p: float = 0.9,
    parse_special: bool = True,
    stream: bool = True,
    cache: KVCache | None = None,
) -> str:
    ids = tokenizer.encode(prompt, parse_special=parse_special)
    dec = StreamDecoder(tokenizer, skip_special=True)
    pieces: list[str] = []
    t0 = time.perf_counter()
    n = 0
    for tid in generate(model, tokenizer, ids, max_tokens, temperature, top_k, top_p, cache):
        n += 1
        chunk = dec.push(tid)
        pieces.append(chunk)
        if stream and chunk:
            sys.stdout.write(chunk)
            sys.stdout.flush()
    tail = dec.flush()
    pieces.append(tail)
    if stream and tail:
        sys.stdout.write(tail)
        sys.stdout.flush()
    elapsed = time.perf_counter() - t0
    if stream:
        tps = n / elapsed if elapsed > 0 else 0.0
        sys.stdout.write(f"\n[{n} tokens, {tps:.2f} tok/s]\n")
        sys.stdout.flush()
    return "".join(pieces)


def generate_batch(
    model: LlamaModel,
    tokenizer: Tokenizer,
    prompts: list[str],
    max_tokens: int = 32,
    temperature: float = 0.0,
    top_k: int = 0,
    top_p: float = 1.0,
    parse_special: bool = True,
) -> list[list[int]]:
    """Padded prefill then per-row decode. Finished rows keep emitting EOS."""
    hp = model.hparams
    id_lists = [tokenizer.encode(p, parse_special=parse_special) for p in prompts]
    pad_id = hp.pad_id
    batch, lengths = _as_batch(id_lists, pad_id)
    bsz = batch.shape[0]
    cache = model.new_cache(batch=bsz)
    logits = model.forward(batch, cache, valid_len=lengths)
    out: list[list[int]] = [[] for _ in range(bsz)]
    done = np.zeros((bsz,), dtype=bool)
    rng = np.random.default_rng()
    for _ in range(max_tokens):
        nxt = np.zeros((bsz, 1), dtype=np.int32)
        for b in range(bsz):
            if done[b]:
                nxt[b, 0] = hp.eos_id
                continue
            tid = sample_token(logits[b], temperature, top_k, top_p, rng)
            if tid in tokenizer.stop_ids:
                nxt[b, 0] = hp.eos_id
                done[b] = True
                continue
            nxt[b, 0] = tid
            out[b].append(tid)
            if tid == hp.eos_id:
                done[b] = True
        if done.all():
            break
        # still-running rows get a real token; finished rows feed EOS (masked by valid_len=0)
        valid = np.where(done, 0, 1).astype(np.int32)
        # rows that are done should not grow the cache
        logits = model.forward(nxt, cache, valid_len=valid)
    return out


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description="Run a completion or a tiny batched demo")
    parser.add_argument("--model", default=str(Path(__file__).resolve().parents[1] / "nanogpt-chat-q8_0.gguf"))
    parser.add_argument("--prompt", default="Hello")
    parser.add_argument("--max-tokens", type=int, default=64)
    parser.add_argument("--temp", type=float, default=0.8)
    parser.add_argument("--top-p", type=float, default=0.9)
    parser.add_argument("--top-k", type=int, default=0)
    parser.add_argument("--batch-demo", action="store_true")
    args = parser.parse_args()

    print(f"loading {args.model}", flush=True)
    from .gguf_loader import load_model

    loaded = load_model(args.model)
    model = LlamaModel(loaded)
    tok = Tokenizer.from_gguf(loaded.gguf, loaded.hparams)
    if args.batch_demo:
        prompts = [
            apply_chat_template([{"role": "user", "content": "Say hi in one word."}]),
            apply_chat_template([{"role": "user", "content": "2+2="}]),
        ]
        print("batched greedy decode:", flush=True)
        outs = generate_batch(model, tok, prompts, max_tokens=args.max_tokens, temperature=0.0)
        for p, ids in zip(prompts, outs):
            print("---")
            print(tok.decode(ids, skip_special=True))
        return
    text = apply_chat_template([{"role": "user", "content": args.prompt}])
    generate_text(model, tok, text, args.max_tokens, args.temp, args.top_k, args.top_p)


if __name__ == "__main__":
    if __package__ in (None, ""):
        sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
        __package__ = "py"
    main()

File: c/generate.c
#include "generate.h"
#include <time.h>

static double monotonic_now(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1e-9;
}

static int is_stop(Tokenizer *tok, int tid) {
    for (int i = 0; i < tok->n_stop; i++)
        if (tok->stop_ids[i] == tid) return 1;
    return 0;
}

void generate_state_init(GenerateState *st, LlamaModel *model, Tokenizer *tok, KVCache *cache,
                         int max_tokens, float temperature, int top_k, float top_p) {
    memset(st, 0, sizeof(*st));
    st->model = model;
    st->tok = tok;
    st->own_cache = cache == NULL;
    st->cache = cache ? cache : llama_model_new_cache(model, 1, model->hparams.n_ctx);
    st->max_tokens = max_tokens;
    st->temperature = temperature;
    st->top_k = top_k;
    st->top_p = top_p;
    rng_seed(&st->rng, (uint64_t)time(NULL) ^ 0xA5A5A5A5ULL);
}

void generate_state_free(GenerateState *st) {
    if (st->own_cache) kvcache_free(st->cache);
}

int generate_start(GenerateState *st, const int *prompt_ids, int n_prompt) {
    st->logits = llama_model_forward(st->model, prompt_ids, 1, n_prompt, st->cache, NULL, NULL);
    st->n_generated = 0;
    st->done = 0;
    return 0;
}

int generate_next(GenerateState *st, int *token_out) {
    if (st->done || st->n_generated >= st->max_tokens) {
        st->done = 1;
        return 1;
    }
    int tid = sample_token(st->logits, st->model->hparams.n_vocab, st->temperature, st->top_k, st->top_p,
                           &st->rng);
    if (is_stop(st->tok, tid)) {
        st->done = 1;
        return 1;
    }
    *token_out = tid;
    st->n_generated++;
    st->logits = llama_model_forward(st->model, &tid, 1, 1, st->cache, NULL, NULL);
    return 0;
}

char *generate_text(LlamaModel *model, Tokenizer *tok, const char *prompt, int max_tokens,
                    float temperature, int top_k, float top_p, int parse_special, int stream,
                    KVCache *cache) {
    IntVec ids;
    intvec_init(&ids);
    tokenizer_encode(tok, prompt, parse_special, &ids);
    GenerateState st;
    generate_state_init(&st, model, tok, cache, max_tokens, temperature, top_k, top_p);
    StreamDecoder dec;
    stream_decoder_init(&dec, tok, 1);
    ByteVec pieces;
    bytevec_init(&pieces);
    double t0 = monotonic_now();
    generate_start(&st, ids.data, ids.n);
    int n = 0;
    int tid;
    while (generate_next(&st, &tid) == 0) {
        n++;
        char *chunk = stream_decoder_push(&dec, tid);
        bytevec_append(&pieces, chunk, (int)strlen(chunk));
        if (stream && chunk[0]) {
            fputs(chunk, stdout);
            fflush(stdout);
        }
        free(chunk);
    }
    char *tail = stream_decoder_flush(&dec);
    bytevec_append(&pieces, tail, (int)strlen(tail));
    if (stream && tail[0]) {
        fputs(tail, stdout);
        fflush(stdout);
    }
    free(tail);
    if (stream) {
        double elapsed = monotonic_now() - t0;
        double tps = elapsed > 0.0 ? (double)n / elapsed : 0.0;
        printf("\n[%d tokens, %.2f tok/s]\n", n, tps);
        fflush(stdout);
    }
    bytevec_push(&pieces, 0);
    stream_decoder_free(&dec);
    generate_state_free(&st);
    intvec_free(&ids);
    return pieces.data;
}

static void as_batch(int **id_lists, const int *lengths, int n, int pad_id, int slen, int *batch) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < slen; j++) batch[i * slen + j] = pad_id;
        for (int j = 0; j < lengths[i]; j++) batch[i * slen + j] = id_lists[i][j];
    }
}

IntVec *generate_batch(LlamaModel *model, Tokenizer *tok, char **prompts, int n_prompts,
                       int max_tokens, float temperature, int top_k, float top_p, int parse_special) {
    LlamaHParams *hp = &model->hparams;
    IntVec *encoded = xcalloc((size_t)n_prompts, sizeof(IntVec));
    int *lengths = xmalloc((size_t)n_prompts * sizeof(int));
    int slen = 0;
    for (int i = 0; i < n_prompts; i++) {
        intvec_init(&encoded[i]);
        tokenizer_encode(tok, prompts[i], parse_special, &encoded[i]);
        lengths[i] = encoded[i].n;
        if (encoded[i].n > slen) slen = encoded[i].n;
    }
    int **id_lists = xmalloc((size_t)n_prompts * sizeof(int *));
    for (int i = 0; i < n_prompts; i++) id_lists[i] = encoded[i].data;
    int *batch = xmalloc((size_t)n_prompts * (size_t)slen * sizeof(int));
    as_batch(id_lists, lengths, n_prompts, hp->pad_id, slen, batch);

    KVCache *cache = llama_model_new_cache(model, n_prompts, hp->n_ctx);
    float *logits = llama_model_forward(model, batch, n_prompts, slen, cache, lengths, NULL);

    IntVec *out = xcalloc((size_t)n_prompts, sizeof(IntVec));
    for (int i = 0; i < n_prompts; i++) intvec_init(&out[i]);
    int *done = xcalloc((size_t)n_prompts, sizeof(int));
    Rng rng;
    rng_seed(&rng, 1);
    int *nxt = xmalloc((size_t)n_prompts * sizeof(int));
    int *valid = xmalloc((size_t)n_prompts * sizeof(int));

    for (int step = 0; step < max_tokens; step++) {
        int all_done = 1;
        for (int b = 0; b < n_prompts; b++) {
            if (done[b]) {
                nxt[b] = hp->eos_id;
                continue;
            }
            all_done = 0;
            int tid = sample_token(logits + (size_t)b * (size_t)hp->n_vocab, hp->n_vocab, temperature,
                                   top_k, top_p, &rng);
            if (is_stop(tok, tid)) {
                nxt[b] = hp->eos_id;
                done[b] = 1;
                continue;
            }
            nxt[b] = tid;
            intvec_push(&out[b], tid);
            if (tid == hp->eos_id) done[b] = 1;
        }
        if (all_done) break;
        for (int b = 0; b < n_prompts; b++) valid[b] = done[b] ? 0 : 1;
        logits = llama_model_forward(model, nxt, n_prompts, 1, cache, valid, NULL);
    }

    kvcache_free(cache);
    for (int i = 0; i < n_prompts; i++) intvec_free(&encoded[i]);
    free(encoded);
    free(lengths);
    free(id_lists);
    free(batch);
    free(done);
    free(nxt);
    free(valid);
    return out;
}
